src/output.rs

```diff
@@ -0,0 +1,56 @@
+use starknet_crypto::{poseidon_hash_many, FieldElement};
+use std::collections::HashMap;
+use std::convert::TryInto;
+
+use crate::parse_raw;
+
+pub struct ExtractOutputResult {
+    pub program_output: Vec<FieldElement>,
+    pub program_output_hash: FieldElement,
+}
+
+pub fn extract_output(input: String) -> anyhow::Result<ExtractOutputResult> {
+    // Parse the input string into a proof structure
+    let proof = parse_raw(&input)?;
+
+    // Retrieve the output segment from the proof
+    let output_segment = proof
+        .public_input
+        .segments
+        .get(2)
+        .ok_or_else(|| anyhow::Error::msg("Output segment not found"))?;
+
+    // Construct a map for the main page elements
+    let mut main_page_map = HashMap::new();
+    for element in &proof.public_input.main_page {
+        let value_bytes = element.value.to_bytes_be();
+        let padded_value = vec![0u8; 32 - value_bytes.len()]
+            .iter()
+            .chain(value_bytes.iter())
+            .copied()
+            .collect::<Vec<u8>>();
+        let field_element = FieldElement::from_bytes_be(
+            &padded_value.try_into().expect("Failed to convert to array"),
+        )
+        .expect("Failed to convert to FieldElement");
+
+        main_page_map.insert(element.address, field_element);
+    }
+
+    // Extract program output using the address range in the output segment
+    let program_output: Vec<FieldElement> = (output_segment.begin_addr..output_segment.stop_ptr)
+        .map(|addr| {
+            *main_page_map
+                .get(&addr)
+                .expect("Address not found in main page map")
+        })
+        .collect();
+
+    // Calculate the Poseidon hash of the program output
+    let program_output_hash = poseidon_hash_many(&program_output);
```

Comment:
So for the program output hash we're still using poseidon, and we only use Pedersen when the provable program needs to hash something, correct?

Reply:
yes, that's correct, it's a standard and verifier does the same